### DISCUSSION TITLE
tsss and true_angular metric

### DISCUSSION THREAD

**Discussion by jlmelville** [lmcinnes/pynndescent#231] [General]
What are the `true_angular` and `tsss` metrics? They were added in the same commit https://github.com/lmcinnes/pynndescent/commit/57d807cb0399ae50856c53ae589113596dd29d89. They have proven resistant to my search engine skills. Is there a reference or something to search for (e.g. what field of study, what does tsss even stand for)?

**Comment by jlmelville** [comment]
To answer my own question (partially) `tsss` is TS-SS https://ieeexplore.ieee.org/document/7474366 (a hybrid of triangle-area similarity and section-area similarity)

**Comment by lmcinnes** [comment]
TS-SS is, as you found, a kind of weird hybrid metric. I had a specific request to add it, and it was easy to implement, so it got thrown in. I'm not sure whether it actually sees any use. ``true_angular`` is another odd case, again from a specific request; cosine distance is one minus the cosine similarity, but potentially one can ask for the actual angle separating two vectors, so this is just the inverse cosine of the cosine similarity. This causes some minor rescaling, especially at either end of the scale (nearly the same, nearly orthogonal), that can apparently have some theoretical benefits for some use cases.  Again it was a case of easy to implement, so just added upon request.

**Comment by jlmelville** [comment]
Thank you!

One other thing: several of these metrics have an alternative with a log2 conversion. This is mentioned at https://pynndescent.readthedocs.io/en/latest/pynndescent_metrics.html#Beware-of-bounded-distances as being to support the beam search during querying with bounded distances, but in the code they are included in a dict `fast_distance_alternatives`. Are they actually faster (due to the extra numba annotations)? Or, if you were only interested in the k-nearest neighbors and not querying an index, would it actually be faster to use the calculations which don't include the log2 transformation?

**Reply by lmcinnes** [reply]
For query they are fast, and there is negligible difference for construction purposes -- where the order of distance computations is all that matters, so any monotonic transform is fine. It was easier to just keep all the faster query alternatives that preserve order in one place and use them for index construction as well. In theory one could separate those out, but I haven't found any benefit in doing so yet, and have just avoided the extra complexity in handling it otherwise.

**Reply by jlmelville** [reply]
I appreciate the details, many thanks.
